src/services/item/plugins/html/h5p/service.ts

```diff
@@ -0,0 +1,89 @@
+import path from 'path';
+import { v4 } from 'uuid';
+
+import { H5PItemExtra, H5PItemType } from '@graasp/sdk';
+
+import { Repositories } from '../../../../../utils/repositories';
+import FileService from '../../../../file/service';
+import { Actor, Member } from '../../../../member/entities/member';
+import { HtmlService } from '../service';
+import { HtmlValidator } from '../validator';
+import { H5P_FILE_DOT_EXTENSION } from './constants';
+import { H5P } from './validation/h5p';
+
+/**
+ * Implementation for the H5P service
+ */
+export class H5PService extends HtmlService {
+  constructor(
+    fileService: FileService,
+    pathPrefix: string,
+    mimetype: string,
+    validator: HtmlValidator,
+  ) {
+    super(fileService, pathPrefix, mimetype, 'h5p', validator);
+  }
+
+  /**
+   * Helper to create H5P extra
+   */
+  buildH5PExtra(contentId: string, filename: string): H5PItemExtra {
+    return {
+      h5p: {
+        contentId,
+        h5pFilePath: this.buildPackagePath(contentId, filename),
+        contentFilePath: this.buildContentPath(contentId),
+      },
+    };
+  }
+
+  /**
+   * Helper to build the local or remote path of the .h5p file
+   */
+  buildH5PPath = (rootPath: string, filename: string) =>
+    path.join(rootPath, `${filename}.${H5P.H5P_FILE_EXTENSION}`);
+
+  /**
+   * Download the H5P file referenced by a given Item
+   */
+  downloadH5P(item: H5PItemType, member: Actor, destinationPath: string) {
+    const h5pPath = item.extra.h5p.h5pFilePath;
+    return this.fileService.download(member, {
+      id: item.id,
+      path: path.join(this.pathPrefix, h5pPath),
+      mimetype: this.mimetype,
+      fileStorage: destinationPath,
+    });
+  }
+
+  async copy(
+    actor: Member,
+    repositories: Repositories,
+    { original: item, copy }: { original: H5PItemType; copy: H5PItemType },
+  ): Promise<void> {
+    const { extra } = item;
+
+    const baseName = path.basename(item.name, H5P_FILE_DOT_EXTENSION);
+    const copySuffix = '-1';
+    const newName = `${baseName}${copySuffix}`;
+
+    const newContentId = v4();
+    const remoteRootPath = this.buildRootPath(this.pathPrefix, newContentId);
+
+    // copy .h5p file
+    await this.fileService.copy(actor, {
+      originalPath: path.join(this.pathPrefix, extra.h5p.h5pFilePath),
+      newFilePath: this.buildH5PPath(remoteRootPath, newName),
+    });
+    // copy content folder
+    await this.fileService.copyFolder(actor, {
+      originalFolderPath: path.join(this.pathPrefix, extra.h5p.contentFilePath),
+      newFolderPath: this.buildContentPath(remoteRootPath),
+    });
```

Comment:
Arguably, copying the assets should be generic as well. But I guess since I wrote above that uploading is not necessarily related to creating an item, then they could be independent as well. Depends on which we chose, but we should stay consistent

Reply:
I also agree, but it's difficult to decoupled this function, everything is tight to h5p.